File: crates/basilica-sdk/src/rl.rs
//! RL Training API client: GRPO post-training on the Basilica platform.
//!
//! Warm GPU clusters, GRPO jobs with custom rewards / datasets / in-cluster
//! LLM-judge, and the declarative manifest surface. The request DTOs here
//! mirror `basilica-api`'s `/rl/*` route DTOs field-for-field (the server
//! uses `deny_unknown_fields`, so the wire shape is the contract); keeping
//! them as shared serde types is what lets the compiler — not a runtime
//! test — catch drift from the server.

use serde::{Deserialize, Serialize};

// Client methods live in `client.rs` (the crate convention: DTOs here, the
// `impl BasilicaClient` beside the private get/post transport helpers).

// ---------------------------------------------------------------------------
// Cluster request DTOs
// ---------------------------------------------------------------------------

/// A trainer/rollout fleet's per-pod GPU shape.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlGpuRequest {
    /// GPU model name, e.g. `H100`, `H200`.
    pub model: String,
    /// GPUs per pod. The trainer total (`count * replicas`) must divide the
    /// GRPO train batch — admission rejects shapes that cannot (valid: 1/2/4/8).
    pub count: u32,
    /// Optional minimum VRAM per GPU in GB. >=140 marks an H200-class fleet,
    /// which the >=16B recipe requires.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_memory_gb: Option<u32>,
}

/// One fleet (trainer or rollout).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlFleetRequest {
    /// Pod replicas (v0 admits exactly 1 for the trainer).
    pub replicas: u32,
    /// Per-pod GPU shape.
    pub gpu: RlGpuRequest,
}

/// Create a warm RL cluster (`POST /rl/clusters`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRlClusterRequest {
    /// Optional cluster name (DNS-1035; generated when omitted).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Base model to pin, e.g. `Qwen/Qwen2.5-7B-Instruct`. Must be on the
    /// platform allowlist; >=16B models require H200-class trainers.
    pub base_model: String,
    /// Trainer fleet (FSDP).
    pub trainer: RlFleetRequest,
    /// Rollout fleet (vLLM).
    pub rollout: RlFleetRequest,
    /// Optional idle-TTL (e.g. `30m`) after which an idle cluster reaps itself.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub idle_ttl: Option<String>,
    /// Bring-your-own relay storage (#1578; server #1574). Omit for the
    /// platform relay — today's behavior, byte-identical on the wire.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub relay: Option<RlRelayRequest>,
    /// Forward-compat catch-all: fields this SDK version doesn't know are
    /// preserved verbatim on the wire (the `body=` escape hatch depends on
    /// this — server-side schema additions must never be silently dropped).
    #[serde(flatten)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

/// Relay storage mode. Serialized lowercase (`byo` | `platform`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RlRelayMode {
    /// User-supplied S3-compatible storage.
    Byo,
    /// The platform-managed relay (the default when the block is omitted).
    Platform,
}

/// BYO relay block (mirrors the server's `RlRelayRequest`, #1574).
///
/// Credentials are WRITE-ONLY on the platform: they become a namespaced
/// secret at create and are never echoed by any response or log. The
/// hand-written `Debug` below redacts them on the CLIENT side too, so an
/// application that traces its requests cannot leak them either.
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlRelayRequest {
    /// Storage mode.
    pub mode: RlRelayMode,
    /// S3-compatible endpoint URL — https with a public-CA cert and a DNS
    /// hostname (IP literals are rejected).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<String>,
    /// Bucket name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bucket: Option<String>,
    /// Region (S3 dialects; R2 uses `auto`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub region: Option<String>,
    /// Organizational prefix inside the bucket; the platform appends
    /// `<cluster-uid>/` — see `effectivePrefix` on the create response.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub base_prefix: Option<String>,
    /// Name of an existing namespaced secret you manage (keys
    /// `access_key_id` / `secret_access_key`). Mutually exclusive with the
    /// inline pair.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credentials_secret: Option<String>,
    /// Inline access key id (write-only; becomes a platform-managed secret).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub access_key_id: Option<String>,
    /// Inline secret access key (write-only).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub secret_access_key: Option<String>,
}

// Redaction policy (deliberate asymmetry): only the two key-material
// fields are redacted. The remaining fields — endpoint, bucket, region,
// basePrefix, and the credentialsSecret NAME — are non-secret storage
// coordinates (the same values appear in the CR spec and cluster status),
// and printing them is what makes a traced request debuggable.
impl std::fmt::Debug for RlRelayRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RlRelayRequest")
            .field("mode", &self.mode)
            .field("endpoint", &self.endpoint)
            .field("bucket", &self.bucket)
            .field("region", &self.region)
            .field("base_prefix", &self.base_prefix)
            .field("credentials_secret", &self.credentials_secret)
            .field(
                "access_key_id",
                &self.access_key_id.as_ref().map(|_| "<redacted>"),
            )
            .field(
                "secret_access_key",
                &self.secret_access_key.as_ref().map(|_| "<redacted>"),
            )
            .finish()
    }
}

/// Rotate a BYO cluster's relay credentials
/// (`POST /rl/clusters/{name}/credentials`, #1577). Applies only to
/// clusters created with the inline pair (platform-managed secret).
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RotateRelayCredentialsRequest {
    /// Replacement access key id (write-only).
    pub access_key_id: String,
    /// Replacement secret access key (write-only).
    pub secret_access_key: String,
}

impl std::fmt::Debug for RotateRelayCredentialsRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("RotateRelayCredentialsRequest")
            .field("access_key_id", &"<redacted>")
            .field("secret_access_key", &"<redacted>")
            .finish()
    }
}

// ---------------------------------------------------------------------------
// Job request DTOs
// ---------------------------------------------------------------------------

/// In-cluster LLM-judge opt-in (WS-G v0.5). The judge is a platform-owned
/// vLLM pod serving an allowlisted open model; the reward reaches it via
/// `ctx["judge"](prompt, ...) -> str`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlJudgeRequest {
    /// Judge model id; omit for the platform default (must be on the judge
    /// allowlist).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
}

/// Custom reward: a user Python scoring function run in the isolated,
/// credential-free, zero-egress executor pod.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlRewardRequest {
    /// Reward ref, `user:<dns-1035-name>`.
    #[serde(rename = "ref")]
    pub reward_ref: String,
    /// Reward source: stdlib Python defining
    /// `reward(prompt, completion, **ctx) -> float` (<=64 KiB).
    pub source: String,
    /// Optional in-cluster LLM-judge.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub judge: Option<RlJudgeRequest>,
}

/// A public Hugging Face dataset source + column mapping.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlHfDatasetSource {
    /// HF dataset id, `org/name` (public datasets only).
    pub repo: String,
    /// Optional HF config (subset) name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub config: Option<String>,
    /// Split to load, e.g. `train`.
    pub split: String,
    /// Column holding the prompt text.
    pub prompt_column: String,
    /// Column handed to the reward as `ground_truth`.
    pub answer_column: String,
}

/// Custom dataset: platform code fetches and renders it (dataset-as-data;
/// no user code touches the data path).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlDatasetRequest {
    /// Dataset ref, `user:<dns-1035-name>`.
    #[serde(rename = "ref")]
    pub dataset_ref: String,
    /// The public HF source + column mapping.
    pub hf: RlHfDatasetSource,
}

/// Create a GRPO training job on a Ready cluster (`POST /rl/jobs`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRlJobRequest {
    /// Name of the warm cluster to bind to.
    pub cluster_ref: String,
    /// Optional job name (generated when omitted).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Algorithm, e.g. `grpo`.
    pub algorithm: String,
    /// Custom reward; omit for the builtin reward.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reward: Option<RlRewardRequest>,
    /// Custom dataset; omit for the builtin GSM8K dataset.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dataset: Option<RlDatasetRequest>,
    /// Training steps (<=3000 on custom datasets).
    pub max_steps: u32,
    /// Optional learning-rate override (string, e.g. `3.0e-6`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub lr: Option<String>,
    /// Forward-compat catch-all: fields this SDK version doesn't know are
    /// preserved verbatim on the wire (the `body=` escape hatch depends on
    /// this — server-side schema additions must never be silently dropped).
    #[serde(flatten)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

/// Declarative manifest: one document that renders a cluster and/or a job
/// (`POST /rl/manifest`). Freeform to mirror the server's document surface.
pub type RlManifestRequest = serde_json::Value;

// ---------------------------------------------------------------------------
// Response DTOs
// ---------------------------------------------------------------------------

/// Response after creating a cluster.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRlClusterResponse {
    /// The cluster's name (its `clusterRef`).
    pub name: String,
    /// Unique identifier.
    pub uid: String,
    /// Lifecycle phase at creation (`Provisioning`).
    pub phase: String,
    /// BYO only: the uid-scoped storage key prefix
    /// (`<basePrefix><cluster-uid>/`) everything the cluster stores lands
    /// under — tighten your IAM grant to it. Absent for platform-managed
    /// storage and on servers predating #1574.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effective_prefix: Option<String>,
}

/// Response after rotating a cluster's storage credentials (#1577).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RotateRlCredentialsResponse {
    /// The cluster whose credentials were rotated.
    pub name: String,
    /// When the rotation was applied (RFC 3339). The relay daemon restarts
    /// with the new key material within seconds of this instant — keep the
    /// OLD key valid until then, then revoke it.
    pub rotated_at: String,
}

/// Cluster status (`GET /rl/clusters/{name}`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlClusterStatusResponse {
    /// `Provisioning` | `Warming` | `Ready` | `Degraded` | `Terminating`.
    pub phase: String,
    /// Whether every fleet pod verified the pinned base model.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model_loaded: Option<bool>,
    /// The bound job's name, if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_job_name: Option<String>,
}

/// Response after creating a job.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRlJobResponse {
    /// The job's name (its identifier for status).
    pub name: String,
    /// Unique identifier.
    pub uid: String,
    /// Lifecycle phase at creation (`Pending`).
    pub phase: String,
    /// sha256 of the admitted reward source (absent for builtin-reward jobs).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reward_sha256: Option<String>,
}

/// Latest training metrics.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlJobMetrics {
    /// Latest training loss.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub loss: Option<f64>,
    /// Latest mean reward.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reward_mean: Option<f64>,
    /// Latest KL divergence.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kl: Option<f64>,
}

/// Job status (`GET /rl/jobs/{name}`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlJobStatusResponse {
    /// `Pending` | `Binding` | `Running` | `Succeeded` | `Failed` | `TimedOut`.
    pub phase: String,
    /// Best-effort latest training step.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub step: Option<i64>,
    /// Latest training metrics.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metrics: Option<RlJobMetrics>,
    /// Object-store location of the trained model (null until bound).
    #[serde(rename = "artifactURI", skip_serializing_if = "Option::is_none")]
    pub artifact_uri: Option<String>,
}

/// Manifest submission result.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RlManifestResponse {
    /// The created cluster, when the manifest carried a `cluster` block.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cluster: Option<CreateRlClusterResponse>,
    /// The created job, when the manifest carried a `job` block.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub job: Option<CreateRlJobResponse>,
}

/// Response after deleting a cluster (`DELETE /rl/clusters/{name}`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteRlClusterResponse {
    /// The deleted cluster's name.
    pub name: String,
}

/// Response after deleting a job (`DELETE /rl/jobs/{name}`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteRlJobResponse {
    /// The deleted job's name.
    pub name: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    // The SDK request DTOs must serialize to the exact wire shape the server's
    // deny_unknown_fields DTOs accept. These pin the field renames and casing
    // that the server contract depends on.
    //
    // Coverage split (deliberate): these are SERIALIZATION tests only. The
    // HTTP-level contract — method, path, auth header, body-on-the-wire,
    // and error mapping for every rl_* call including credential rotation —
    // is pinned by the Python suite's recorded-HTTP harness
    // (crates/basilica-sdk-python/tests/test_rl_client.py), which exercises
    // the COMPILED core transport end to end. A Rust-side HTTP harness
    // would duplicate that coverage one layer lower.
    #[test]
    fn job_request_wire_shape() {
        let req = CreateRlJobRequest {
            cluster_ref: "my-pool".into(),
            name: None,
            algorithm: "grpo".into(),
            reward: Some(RlRewardRequest {
                reward_ref: "user:my-reward".into(),
                source: "def reward(p, c, **k):\n    return 1.0\n".into(),
                judge: Some(RlJudgeRequest { model: None }),
            }),
            dataset: Some(RlDatasetRequest {
                dataset_ref: "user:my-data".into(),
                hf: RlHfDatasetSource {
                    repo: "openai/gsm8k".into(),
                    config: Some("main".into()),
                    split: "train".into(),
                    prompt_column: "question".into(),
                    answer_column: "answer".into(),
                },
            }),
            max_steps: 50,
            lr: None,
            extra: Default::default(),
        };
        let v = serde_json::to_value(&req).unwrap();
        assert_eq!(v["clusterRef"], "my-pool");
        assert_eq!(v["maxSteps"], 50);
        // the serde renames: nested identity fields are `ref`
        assert_eq!(v["reward"]["ref"], "user:my-reward");
        assert_eq!(v["dataset"]["ref"], "user:my-data");
        assert_eq!(v["dataset"]["hf"]["promptColumn"], "question");
        // judge with no model serializes as an empty object (not null)
        assert!(v["reward"]["judge"].is_object());
        assert_eq!(v["reward"]["judge"].as_object().unwrap().len(), 0);
        // None optionals are ABSENT (deny_unknown_fields tolerates absence,
        // but a null in a non-Option server slot would 400)
        assert!(v.get("name").is_none());
        assert!(v.get("lr").is_none());
    }

    #[test]
    fn delete_response_wire_shape() {
        // The server serializes camelCase; both delete responses carry only
        // `name`. Pin the deserialization so a server-side field rename is
        // caught here, not by a user.
        let c: DeleteRlClusterResponse = serde_json::from_str(r#"{"name":"my-pool"}"#).unwrap();
        assert_eq!(c.name, "my-pool");
        let j: DeleteRlJobResponse = serde_json::from_str(r#"{"name":"my-pool-job"}"#).unwrap();
        assert_eq!(j.name, "my-pool-job");
    }

    #[test]
    fn cluster_request_wire_shape() {
        let req = CreateRlClusterRequest {
            name: Some("my-pool".into()),
            base_model: "Qwen/Qwen2.5-7B-Instruct".into(),
            trainer: RlFleetRequest {
                replicas: 1,
                gpu: RlGpuRequest {
                    model: "H100".into(),
                    count: 4,
                    min_memory_gb: None,
                },
            },
            rollout: RlFleetRequest {
                replicas: 1,
                gpu: RlGpuRequest {
                    model: "H100".into(),
                    count: 4,
                    min_memory_gb: None,
                },
            },
            idle_ttl: Some("30m".into()),
            relay: None,
            extra: Default::default(),
        };
        let v = serde_json::to_value(&req).unwrap();
        assert_eq!(v["baseModel"], "Qwen/Qwen2.5-7B-Instruct");
        assert_eq!(v["trainer"]["gpu"]["count"], 4);
        assert_eq!(v["idleTtl"], "30m");
        assert!(v["trainer"]["gpu"].get("minMemoryGb").is_none());
        // No relay block => byte-identical to the pre-#1578 wire shape.
        assert!(v.get("relay").is_none());
    }

    #[test]
    fn relay_request_wire_shape() {
        let relay = RlRelayRequest {
            mode: RlRelayMode::Byo,
            endpoint: Some("https://acc.r2.cloudflarestorage.com".into()),
            bucket: Some("my-weights".into()),
            region: None,
            base_prefix: Some("teams/rl/".into()),
            credentials_secret: None,
            access_key_id: Some("AK".into()),
            secret_access_key: Some("SK".into()),
        };
        let v = serde_json::to_value(&relay).unwrap();
        // The server's deny_unknown_fields DTO expects exactly these names.
        assert_eq!(v["mode"], "byo");
        assert_eq!(v["basePrefix"], "teams/rl/");
        assert_eq!(v["accessKeyId"], "AK");
        assert_eq!(v["secretAccessKey"], "SK");
        assert!(v.get("region").is_none());
        assert!(v.get("credentialsSecret").is_none());
    }

    #[test]
    fn relay_debug_redacts_credentials() {
        let relay = RlRelayRequest {
            mode: RlRelayMode::Byo,
            endpoint: None,
            bucket: None,
            region: None,
            base_prefix: None,
            credentials_secret: None,
            access_key_id: Some("SUPERSECRETAK".into()),
            secret_access_key: Some("SUPERSECRETSK".into()),
        };
        let dbg = format!("{relay:?}");
        assert!(!dbg.contains("SUPERSECRET"), "Debug must redact: {dbg}");
        assert!(dbg.contains("<redacted>"));
        let rot = RotateRelayCredentialsRequest {
            access_key_id: "SUPERSECRETAK".into(),
            secret_access_key: "SUPERSECRETSK".into(),
        };
        let dbg = format!("{rot:?}");
        assert!(!dbg.contains("SUPERSECRET"), "Debug must redact: {dbg}");
    }

    #[test]
    fn cluster_response_effective_prefix_is_optional() {
        // Old servers (pre-#1574) omit it; BYO creates carry it.
        let old: CreateRlClusterResponse =
            serde_json::from_str(r#"{"name":"p","uid":"u","phase":"Provisioning"}"#).unwrap();
        assert_eq!(old.effective_prefix, None);
        let byo: CreateRlClusterResponse = serde_json::from_str(
            r#"{"name":"p","uid":"u","phase":"Provisioning","effectivePrefix":"teams/rl/u/"}"#,
        )
        .unwrap();
        assert_eq!(byo.effective_prefix.as_deref(), Some("teams/rl/u/"));
    }

    #[test]
    fn rotation_wire_shapes() {
        let req = RotateRelayCredentialsRequest {
            access_key_id: "AK".into(),
            secret_access_key: "SK".into(),
        };
        let v = serde_json::to_value(&req).unwrap();
        assert_eq!(v["accessKeyId"], "AK");
        assert_eq!(v["secretAccessKey"], "SK");
        let resp: RotateRlCredentialsResponse =
            serde_json::from_str(r#"{"name":"p","rotatedAt":"2026-08-31T12:00:00+00:00"}"#)
                .unwrap();
        assert_eq!(resp.name, "p");
        assert!(resp.rotated_at.starts_with("2026-08-31"));
    }

    #[test]
    fn job_status_parses_artifact_uri_rename() {
        let body = r#"{"phase":"Succeeded","step":50,"artifactURI":"s3://x/uid"}"#;
        let s: RlJobStatusResponse = serde_json::from_str(body).unwrap();
        assert_eq!(s.phase, "Succeeded");
        assert_eq!(s.artifact_uri.as_deref(), Some("s3://x/uid"));
        assert_eq!(s.step, Some(50));
    }
}

File: crates/basilica-sdk-python/python/basilica/rl.py
"""RL training namespace: GRPO post-training on the Basilica RL Training API.

    >>> from basilica import BasilicaClient
    >>> client = BasilicaClient()               # BASILICA_API_TOKEN / CLI login
    >>> client.rl.create_cluster(
    ...     name="my-pool",
    ...     base_model="Qwen/Qwen2.5-7B-Instruct",
    ...     gpu_model="H100",
    ... )
    >>> client.rl.wait_cluster("my-pool")
    >>> job = client.rl.create_job(
    ...     cluster="my-pool", max_steps=50,
    ...     reward_name="my-reward", reward_source=REWARD_PY,
    ...     dataset_name="my-data", dataset_repo="openai/gsm8k",
    ...     dataset_config="main", dataset_split="train",
    ...     prompt_column="question", answer_column="answer",
    ... )
    >>> final = client.rl.wait_job(job["name"])  # {phase, step, metrics, artifactURI}

THIN WRAPPER over the compiled core: this module builds the ergonomic
kwargs into wire dicts and hands them to the Rust binding's ``rl_*``
methods, which serde-validate against the core's typed DTOs
(``basilica_sdk::rl`` — the compile-time-shared contract with the server)
and send through the core transport. That inherits the full auth chain
(explicit key, BASILICA_API_TOKEN, and the CLI-login token fallback) and
the core's error mapping: non-2xx surfaces as ValueError (bad request),
PermissionError (authz), ConnectionError (transport), KeyError
(not found), or RuntimeError (server error), each carrying the server's
message verbatim.

The ``body=`` escape hatch on every create call sends a raw dict; unknown
fields survive the typed round-trip verbatim (serde-flatten catch-alls in
the core DTOs), so server-side schema additions never strand you on an SDK
release.
"""

from __future__ import annotations

import json
import time
from typing import Any, Optional

_TERMINAL_JOB_PHASES = frozenset({"Succeeded", "Failed", "TimedOut"})
# Degraded is deliberately NOT here: a cluster degrades on transient fleet
# unhealth (pod restart, node blip) and can recover to Ready; only
# Terminating can never become Ready again.
_DEAD_CLUSTER_PHASES = frozenset({"Terminating"})
# A single LB 502 or connection reset must not abort a multi-hour wait;
# this many CONSECUTIVE poll failures (reset on any success) give up.
_POLL_FAILURE_BUDGET = 5


def _drop_none(d: dict) -> dict:
    return {k: v for k, v in d.items() if v is not None}


class RlNamespace:
    """The ``client.rl`` surface. Constructed by BasilicaClient; hold no
    credentials here — the compiled core owns auth and transport."""

    def __init__(self, core: Any):
        self._core = core

    # -- clusters ----------------------------------------------------------

    def create_cluster(
        self,
        *,
        base_model: Optional[str] = None,
        gpu_model: Optional[str] = None,
        trainer_gpus: int = 4,
        rollout_gpus: int = 4,
        name: Optional[str] = None,
        min_memory_gb: Optional[int] = None,
        idle_ttl: Optional[str] = None,
        relay: Optional[dict] = None,
        body: Optional[dict] = None,
    ) -> dict:
        """POST /rl/clusters — a warm trainer+rollout GPU pool.

        Certified shapes: 4+4 (H100 for <16B models, H200-class for >=16B —
        admission rejects bad pairings with an actionable message).

        ``relay`` selects bring-your-own storage (wire-shaped dict; the same
        dict works inside a manifest's ``cluster`` block)::

            relay={
                "mode": "byo",
                "endpoint": "https://<acct>.r2.cloudflarestorage.com",
                "bucket": "my-weights",
                "basePrefix": "teams/rl/",          # optional
                "accessKeyId": "...",               # write-only: becomes a
                "secretAccessKey": "...",           #   secret, never echoed
                # or instead of the pair: "credentialsSecret": "my-secret"
            }

        The response then carries ``effectivePrefix`` — the uid-scoped key
        prefix all cluster data lands under; tighten your IAM grant to it.
        Omit ``relay`` for platform-managed storage (today's behavior).
        ``body`` replaces the built request entirely (raw wire dict, escape
        hatch) — no other kwargs are consulted when it is given.
        """
        if body is None:
            if base_model is None or gpu_model is None:
                raise ValueError(
                    "base_model and gpu_model are required (unless a raw body= is given)"
                )

            def fleet(count: int) -> dict:
                return {
                    "replicas": 1,
                    "gpu": _drop_none(
                        {"model": gpu_model, "count": count, "minMemoryGb": min_memory_gb}
                    ),
                }

            body = _drop_none(
                {
                    "name": name,
                    "baseModel": base_model,
                    "trainer": fleet(trainer_gpus),
                    "rollout": fleet(rollout_gpus),
                    "idleTtl": idle_ttl,
                    "relay": relay,
                }
            )
        return json.loads(self._core.rl_create_cluster(json.dumps(body)))

    def rotate_credentials(
        self, name: str, *, access_key_id: str, secret_access_key: str
    ) -> dict:
        """POST /rl/clusters/{name}/credentials — rotate a BYO cluster's
        storage credentials (only clusters created with the inline pair;
        clusters using ``credentialsSecret`` update their own secret).

        Sequencing: create the NEW key at your provider first (both keys
        valid), call this, then revoke the OLD key after the returned
        ``rotatedAt`` plus a few seconds — the relay daemon restarts onto
        the new material in that window; in-flight transfers finish on the
        old key. Revoking first fails the running job with
        ``RelayAuthFailed`` (recoverable, but wastes the retry window)."""
        req = {"accessKeyId": access_key_id, "secretAccessKey": secret_access_key}
        return json.loads(
            self._core.rl_rotate_cluster_credentials(name, json.dumps(req))
        )

    def get_cluster(self, name: str) -> dict:
        return json.loads(self._core.rl_get_cluster(name))

    def delete_cluster(self, name: str) -> dict:
        """Delete a cluster. Refused (ValueError) while a job is active —
        the error names the blocking job; delete the job first (that IS the
        cancel path). Deleting the namespace's last cluster also tears down
        its RL prerequisites server-side."""
        return json.loads(self._core.rl_delete_cluster(name))

    def delete_job(self, name: str) -> dict:
        """Delete a job — valid in any phase. Deleting a running job IS the
        cancel path (pods are torn down by the operator's stop ladder)."""
        return json.loads(self._core.rl_delete_job(name))

    def wait_cluster(
        self, name: str, timeout_s: float = 1800.0, poll_s: float = 15.0
    ) -> dict:
        """Poll until phase == Ready. Raises RuntimeError immediately on
        Terminating (it can never become Ready — waiting out the full
        timeout would hide the failure), TimeoutError on the deadline.
        Degraded keeps polling: fleets recover from transient unhealth."""
        deadline = time.monotonic() + timeout_s
        failures = 0
        while True:
            try:
                cluster = self.get_cluster(name)
            except (ConnectionError, RuntimeError):
                failures += 1
                if failures >= _POLL_FAILURE_BUDGET:
                    raise
                time.sleep(poll_s)
                continue
            failures = 0
            phase = cluster.get("phase")
            if phase == "Ready":
                return cluster
            if phase in _DEAD_CLUSTER_PHASES:
                raise RuntimeError(f"cluster {name!r} entered {phase}: {cluster}")
            if time.monotonic() >= deadline:
                raise TimeoutError(
                    f"cluster {name!r} not Ready after {timeout_s}s "
                    f"(last phase: {phase!r})"
                )
            time.sleep(poll_s)

    # -- jobs --------------------------------------------------------------

    def create_job(
        self,
        *,
        cluster: Optional[str] = None,
        max_steps: Optional[int] = None,
        name: Optional[str] = None,
        algorithm: str = "grpo",
        # custom reward (user:<name> + inline source); omit for the builtin
        reward_name: Optional[str] = None,
        reward_source: Optional[str] = None,
        judge: bool = False,
        judge_model: Optional[str] = None,
        # custom dataset (public HF repo + column mapping); omit for builtin
        dataset_name: Optional[str] = None,
        dataset_repo: Optional[str] = None,
        dataset_split: Optional[str] = None,
        dataset_config: Optional[str] = None,
        prompt_column: Optional[str] = None,
        answer_column: Optional[str] = None,
        lr: Optional[str] = None,
        body: Optional[dict] = None,
    ) -> dict:
        """POST /rl/jobs — a GRPO training job on a Ready cluster.

        The reward is any deterministic stdlib-Python
        ``reward(prompt, completion, **ctx) -> float``; it runs in an
        isolated credential-free pod. ``judge=True`` exposes
        ``ctx["judge"](prompt)`` backed by an in-cluster judge model
        (requires a custom reward). ``body`` replaces the built request
        entirely — no other kwargs are consulted when it is given.

        Orphan kwargs raise: a ``reward_source`` without ``reward_name`` (or
        dataset fields without ``dataset_name``) would otherwise be silently
        dropped and the BUILTIN reward/dataset would run on a paid GPU job.
        """
        if body is None:
            if cluster is None or max_steps is None:
                raise ValueError(
                    "cluster and max_steps are required (unless a raw body= is given)"
                )
            reward = None
            if reward_name is not None:
                if reward_source is None:
                    raise ValueError("reward_source is required with reward_name")
                reward = {"ref": f"user:{reward_name}", "source": reward_source}
                if judge or judge_model:
                    reward["judge"] = _drop_none({"model": judge_model})
            elif reward_source is not None:
                raise ValueError("reward_name is required with reward_source")
            elif judge or judge_model:
                raise ValueError(
                    "judge requires a custom reward (it is called from your reward code)"
                )
            dataset = None
            if dataset_name is not None:
                dataset = {
                    "ref": f"user:{dataset_name}",
                    "hf": _drop_none(
                        {
                            "repo": dataset_repo,
                            "config": dataset_config,
                            "split": dataset_split,
                            "promptColumn": prompt_column,
                            "answerColumn": answer_column,
                        }
                    ),
                }
            elif any(
                v is not None
                for v in (
                    dataset_repo,
                    dataset_config,
                    dataset_split,
                    prompt_column,
                    answer_column,
                )
            ):
                raise ValueError("dataset_name is required with dataset fields")
            body = _drop_none(
                {
                    "clusterRef": cluster,
                    "name": name,
                    "algorithm": algorithm,
                    "maxSteps": max_steps,
                    "reward": reward,
                    "dataset": dataset,
                    "lr": lr,
                }
            )
        return json.loads(self._core.rl_create_job(json.dumps(body)))

    def get_job(self, name: str) -> dict:
        return json.loads(self._core.rl_get_job(name))

    def wait_job(
        self, name: str, timeout_s: float = 6 * 3600.0, poll_s: float = 30.0
    ) -> dict:
        """Poll until the job is terminal (Succeeded/Failed/TimedOut) and
        return the final document either way — check ``phase`` yourself;
        raising on Failed would hide the failure detail behind an
        exception. Transient poll errors are tolerated up to
        ``_POLL_FAILURE_BUDGET`` consecutive failures — a single LB blip
        must not abort a multi-hour wait."""
        deadline = time.monotonic() + timeout_s
        failures = 0
        while True:
            try:
                job = self.get_job(name)
            except (ConnectionError, RuntimeError):
                failures += 1
                if failures >= _POLL_FAILURE_BUDGET:
                    raise
                time.sleep(poll_s)
                continue
            failures = 0
            if job.get("phase") in _TERMINAL_JOB_PHASES:
                return job
            if time.monotonic() >= deadline:
                raise TimeoutError(
                    f"job {name!r} not terminal after {timeout_s}s "
                    f"(last phase: {job.get('phase')!r})"
                )
            time.sleep(poll_s)

    # -- manifest (declarative: one document -> cluster and/or job) --------

    def submit_manifest(self, manifest: dict) -> dict:
        return json.loads(self._core.rl_submit_manifest(json.dumps(manifest)))
